docs/proposals/authentication-filter.md

```diff
@@ -22,3 +22,1101 @@ This new filter should eventually expose all forms of authentication available t
+// +kubebuilder:validation:XValidation:message="when spec.jwt is set, type must be 'JWT'",rule="self.jwt != null ? self.type == 'JWT' : true"
+type AuthenticationFilterSpec struct {
+  // Type selects the authentication mechanism.
+  Type AuthType `json:"type"`
+
+  // Basic configures HTTP Basic Authentication.
+  // Required when Type == Basic.
+  //
+  // +optional
+  Basic *BasicAuth `json:"basic,omitempty"`
+
+  // JWT configures JSON Web Token authentication (NGINX Plus).
+  // Required when Type == JWT.
+  //
+  // +optional
+  JWT *JWTAuth `json:"jwt,omitempty"`
+}
+
+// AuthType defines the authentication mechanism.
+// +kubebuilder:validation:Enum=Basic;JWT
+type AuthType string
+
+const (
+  AuthTypeBasic AuthType = "Basic"
+  AuthTypeJWT   AuthType = "JWT"
+)
+
+// BasicAuth configures HTTP Basic Authentication.
+type BasicAuth struct {
+  // SecretRef allows referencing a Secret in the same or different namespace.
+  // When namespace is set and differs from the filter's namespace, a ReferenceGrant in the target namespace is required.
+  //
+  // +optional
+  SecretRef *SecretObjectReference `json:"secretRef,omitempty"`
+
+  // Realm used by NGINX `auth_basic` directive.
+  // https://nginx.org/en/docs/http/ngx_http_auth_basic_module.html#auth_basic
+  // Also configures "realm="<realm_value>" in WWW-Authenticate header in error page location.
+  //
+  // +optional
+  Realm *string `json:"realm,omitempty"`
+
+  // OnFailure customizes the 401 response for failed authentication.
+  //
+  // +optional
+  OnFailure *AuthFailureResponse `json:"onFailure,omitempty"`
+}
+
+// JWTKeyMode selects where JWT keys come from.
+// +kubebuilder:validation:Enum=File;Remote
+type JWTKeyMode string
+
+const (
+  JWTKeyModeFile   JWTKeyMode = "File"
+  JWTKeyModeRemote JWTKeyMode = "Remote"
+)
+
+// JWTAuth configures JWT-based authentication (NGINX Plus).
+// +kubebuilder:validation:XValidation:message="mode 'File' requires file set and remote unset",rule="self.mode == 'File' ? self.file != null && self.remote == null : true"
+// +kubebuilder:validation:XValidation:message="mode 'Remote' requires remote set and file unset",rule="self.mode == 'Remote' ? self.remote != null && self.file == null : true"
+// +kubebuilder:validation:XValidation:message="when file is set, mode must be 'File'",rule="self.file != null ? self.mode == 'File' : true"
+// +kubebuilder:validation:XValidation:message="when remote is set, mode must be 'Remote'",rule="self.remote != null ? self.mode == 'Remote' : true"
+type JWTAuth struct {
+  // Realm used by NGINX `auth_jwt` directive
+  // https://nginx.org/en/docs/http/ngx_http_auth_jwt_module.html#auth_jwt
+  // Configures "realm="<realm_value>" in WWW-Authenticate header in error page location.
+  //
+  // +optional
+  // +kubebuilder:default="Restricted"
+  Realm *string `json:"realm,omitempty"`
+
+  // Mode selects how JWT keys are provided: local file or remote JWKS.
+  Mode JWTKeyMode `json:"mode,omitempty"`
```

Comment:
I think this should probably be named KeyMode to be more specific.

Reply:
Sorry, when I commented this I meant the field name, not the type. The type can stay as JWTKeyMode

Reply:
Same with the other fields that I left similar comments on.